#' function to automatically retrieve sequences from a databank based on
#'   accession numbers
#'
#' this is a function to retrieve sequences of DNA, RNA or protein from
#'   databases structured under ACNUC and located on the web, the retrieval is
#'   limited to accession numbers as selection criteria, remote access to ACNUC
#'   databases works by opening a socket connection on a port (for example on port
#'   number 5558 at pbil.univ-lyon1.fr) and by communicating on this socket
#'   following the protocol described \href{http://doua.prabi.fr/databases/acnuc/remote_acnuc.html}{here}
#'
#'
#' @param seqnames list of accession numbers as character strings of the
#'   sequences to be retrieved
#' @param acnucdb character string with the name of the ACNUC database to be
#'   searched. Use \code{\link[seqinr]{choosebank}} without arguments to see a
#'   list of available databases
#'
#' @return this function returns a named list with the sequences matching the
#'   sequence names of class SeqAcnucWeb retrieved by
#'   \code{\link[seqinr]{query}} based on the provided accession numbers the
#'   name of each list item is composed of the accession number and the length
#'   of the sequence
#'
#' @examples
#' seqs <- retrieve_seqs(c('P06747', 'P0C569', 'O56773', 'Q5VKP1'), 'swissprot')
#' \dontrun{viruses_annot <- read.csv(system.file('extdata', 'CoV_genomes_annot.csv', package='firstPackage'))}
#' \dontrun{accessions <- viruses_annot$accession}
#' \dontrun{seqs <- retrieve_seqs(accessions, 'genbank')}
#' \dontrun{str(seqs)}
#'
#' @importFrom seqinr choosebank query getSequence
#'
#' @export
retrieve_seqs <- function(seqnames, acnucdb){
                          myseqs <- list()
                          require('seqinr')
                          choosebank(acnucdb)
                          for(i in 1:length(seqnames)){
                              seqname <- seqnames[i]
                              print(paste('retrieving sequence', seqname, '...'))
                              query <- paste('AC=', seqname, sep='')
                              query2 <- query('query2', `query`)
                              seq <- getSequence(query2$req[[1]])
                              myseqs[[i]] <- seq
                              name <- paste(seqname, 'length', length(seq), sep=' ')
                              names(myseqs)[[i]] <- name
                              }
                          closebank()
                          return(myseqs)
                          }

#' function to print an ordered sequences alignment in chunks of defined size
#'
#' this is a function that takes an alignment in phylip format made by an
#'   alignment software like clustal omega and print it to the console in an
#'   ordered display of defined sequence width the function calculates the number
#'   of letters printed so far and print this number together with the name of
#'   each sequence at the end of each chunk
#'
#' @param alignment object of type alignment with the aligned sequences to be
#'   display. A file in format phylip is generated by an alignment software from
#'   a .fasta file containing the sequences and loaded to the R environment
#' @param chunksize number of sequence letters to be printed in each chunk of
#'   display
#'
#' @return this function print the sequence alignment to the console in an
#'   ordered fashion
#'
#' @examples
#' virusaln <- read.alignment(file=system.file('extdata', 'phosphoproteins.phy', package='firstPackage'), format='phylip')
#' print_alignment(virusaln, 80)
#'
#' @importFrom seqinr read.alignment
#' @importFrom Biostrings countPattern
#'
#' @export
print_alignment <- function(alignment, chunksize=60){
                            require('Biostrings')
                            numseqs <- alignment$nb
                            alignmentlen <- nchar(alignment$seq[[1]])
                            starts <- seq(1, alignmentlen, by=chunksize)
                            n <- length(starts)
                            aln <- vector()
                            lettersprinted <- vector()
                            for(j in 1:numseqs){
                                alignmentj <- alignment$seq[[j]]
                                aln[j] <- alignmentj
                                lettersprinted[j] <- 0
                                }
                            for(i in 1:n){
                                for(j in 1:numseqs){
                                    alnj <- aln[j]
                                    chunkseqjaln <- substring(alnj, starts[i], starts[i]+chunksize-1)
                                    chunkseqjaln <- toupper(chunkseqjaln)
                                    gapsj <- countPattern('-', chunkseqjaln)
                                    lettersprinted[j] <- lettersprinted[j]+chunksize-gapsj
                                    print(paste(chunkseqjaln, lettersprinted[j], alignment$nam[j]), quote=FALSE)
                                    }
                                print(paste(''), quote=FALSE)
                                }
                            }
